Replace lambdas with function calls

Solomon1732 opened the pull request:
Replace lambdas with function calls. This was done either by extracting a lambda into a function or method, or by replacing it with a library function call. Also some minor code quality changes.

This PR fixes https://github.com/Tribler/tribler/issues/7432

Solomon1732 commented:
This pull request is instead of #7441 which I messed up.

Solomon1732 commented:
I didn't see the test result for "codecov/patch", so I thought it was fine to convert to "ready for review." Now I see I still need to fix it, so I converted it back to "draft".

drew2a commented on src/tribler/core/components/tunnel/community/dispatcher.py:
```diff
@@ -120,8 +126,13 @@ def select_circuit(self, connection, request):
             self.cid_to_con[circuit.circuit_id] = connection
-            circuit.ready.add_done_callback(lambda f, c=connection.udp_connection, r=request:
-                                            self.on_socks5_udp_data(c, r) if f.result() else None)
+            circuit.ready.add_done_callback(
+                functools.partial(
+                    self._add_data_if_result,
+                    connection=connection.udp_connection,
+                    request=request
+                )
+            )
```
Could you please consider simplifying it, something along the lines of the following:
```suggestion
 
            def done_callback(f, c=connection.udp_connection, r=request):
                return self.on_socks5_udp_data(c, r) if f.result() else None

            circuit.ready.add_done_callback(done_callback)
```

Solomon1732 replied:
Will do.

drew2a commented on src/tribler/core/components/tunnel/community/tunnel_community.py:
```diff
@@ -503,7 +504,7 @@ def monitor_downloads(self, dslist):
                 if old_state != DownloadStatus.METADATA or new_state != DownloadStatus.DOWNLOADING:
                     self.join_swarm(info_hash, hops[info_hash], seeding=new_state == DownloadStatus.SEEDING,
-                                    callback=lambda addr, ih=info_hash: self.on_e2e_finished(addr, ih))
+                                    callback=functools.partial(self.on_e2e_finished, info_hash=info_hash))
```
The new version appears to be more complex. Please contemplate retaining the lambda, or consider substituting it with a function, such as:

```python
                    def callback(addr, ih=info_hash):
                        return self.on_e2e_finished(addr, ih)

                    self.join_swarm(info_hash, hops[info_hash], seeding=new_state == DownloadStatus.SEEDING,
                                    callback=callback)
```

Solomon1732 replied:
Can you please explain why it appears more complex? It does the same thing. And afacit, it has the same information😅.

drew2a replied:
I surmise that a larger number of code readers are accustomed to lambdas rather than `functools.partial`. Hence, they might find the lambda version more readable. Since there are no significant advantages of using `functools.partial` over lambdas in this particular context, I believe the lambda version is preferable. Does that make sense?

Solomon1732 replied:
Yes. I'll reverse the change.

drew2a reviewed (changes requested):
Thank you for the tremendous effort you've put into this work. The quality of the Tribler codebase has improved with these changes. However, I have some comments which I've added for your consideration.

Solomon1732 commented on src/tribler/core/components/gigachannel/community/gigachannel_community.py:
```diff
@@ -50,7 +51,7 @@ def add(self, peer: Peer, channel_pk: bytes, channel_id: int):
 
         if len(channel_peers) > self.max_peers_per_channel:
-            removed_peer = min(channel_peers, key=lambda x: x.last_response)
+            removed_peer = min(channel_peers, key=operator.attrgetter("last_response"))
```
I admit I'm not familiar with the term "backreferencing" when not used in the context of regex. My reply is thus limited by this knowledge gap.

My reply:
Afaict, they produce the same result. Here's an example:
```python
import operator

class Strng:
    def __init__(self, strng: str):
        self.strng: str = strng

example_text = """Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum."""
example_text = [Strng(s) for s in example_text .split()]

get_str = operator.attrgetter("strng")
get_str_lambda = lambda x: x.strng

print(get_str(example_text[0])
print(get_str_lambda(example_text[0])
print(get_str(example_text[2])
print(get_str_lambda(example_text[2])
```

The difference is one generates a lambda. It is represented with something like `<function <lambda> at 0x0000011B0C600040>`. The other an `operator` module object. This is represented with the likes of `operator.attrgetter('strng')`. When used as a function, they both produce the same result.

Here's the description from the documentation for [`operator`](https://docs.python.org/3/library/operator.html):
> The operator module exports a set of efficient functions corresponding to the intrinsic operators of Python. For example, operator.add(x, y) is equivalent to the expression x+y.

As far as I understand, my proposed change isn't more limited than the already existing lambda. I simply switched repeatedly defined lambdas with a call to a built-in standard library module.

drew2a replied:
Perhaps I'm misusing the term `backreferencing`. 

What I mean by `backreference` is the ability of an IDE to find all usages of a class/function, etc.
In this particular case, I'm referring to the "Find Usage" feature of PyCharm.
https://www.jetbrains.com/help/pycharm/finding-usages.html

I'm operating under the assumption that most developers nowadays use an IDE to write code, and I see a benefit in writing code that can be processed effectively by an IDE.

I'll try to illustrate my point with some examples.

Imagine you have a class definition (in file 1):

```python
class Peer:
    def __init__(self, last_response=42):
        self.last_response = last_response

    def __repr__(self):
        return str(self.last_response)
```

In a different part of the code, a list instance might be created (in file 2):

```python
channel_peers = [Peer(last_response=3), Peer(last_response=2), Peer(last_response=1), ]
```

And in yet another place, without direct import, the following operations could be performed (in file 3):

```python
# the following two lines are logically similar:
print(min(channel_peers, key=lambda b: b.last_response))  # 1
print(min(channel_peers, key=operator.attrgetter('last_response')))  # 1
```

Suppose a developer wants to rename the `last_response` attribute to, say, `updated`. By using an IDE, it's relatively easy to find all references to the property when it's used in the form `key=lambda b: b.last_response`. However, to find all references to the property when it's used in the form `operator.attrgetter('last_response')`, you would need to manually search for `last_response`.

Both versions are acceptable, but from my perspective, the version with the lambda is a bit easier to maintain.

Solomon1732 replied:
Ohhh. I'll revert this change 😃.

Solomon1732 replied:
This lambda was written twice, so I opted to extract both instances to a static method (with a single underscore to mark it as internal).

Solomon1732 commented on doc/conf.py:
```diff
@@ -49,8 +57,7 @@
 
     add_endpoint = RootEndpoint.add_endpoint
-    RootEndpoint.add_endpoint = lambda self, path, ep: add_endpoint(self, path, ep) \
-        if path not in ['/ipv8', '/market', '/wallets'] else None
+    RootEndpoint.add_endpoint = _add_root_endpoint(add_endpoint=add_endpoint)
```
I'll change it accordingly.

Solomon1732 commented on src/tribler/core/components/gigachannel/community/gigachannel_community.py:
```diff
@@ -67,7 +68,10 @@ def remove_peer(self, peer):
         id_tuple = (channel_pk, channel_id)
         channel_peers = self._channels_dict.get(id_tuple, [])
-        return sorted(channel_peers, key=lambda x: x.last_response, reverse=True)[0:limit]
+        last_seen_peers = (
+            sorted(channel_peers, key=operator.attrgetter("last_response"), reverse=True)
+        )
+        return last_seen_peers[0:limit]
```
My response is the same as here https://github.com/Tribler/tribler/pull/7466#discussion_r1248039635

Solomon1732 replied:
Ditto.

Solomon1732 commented on src/tribler/core/components/libtorrent/download_manager/download_manager.py:
```diff
@@ -363,7 +363,8 @@ def set_max_connections(self, conns, hops=None):
         # Rate conversion due to the fact that we had a different system with Swift
         # and the old python BitTorrent core: unlimited == 0, stop == -1, else rate in kbytes
-        libtorrent_rate = int(-1 if rate == 0 else (1 if rate == -1 else rate * 1024))
+        torrent_rate = {0: -1, -1: 1}
```
I do plan to add a test before marking it as ready. Like I said here https://github.com/Tribler/tribler/pull/7466#issuecomment-1614588631, I converted it from "draft" to "ready" by accident 😅.

I changed it to use a dictionary first to make it clearer, and second, to avoid branching. A series of `if` statements like you wrote is clearer than what I wrote, but in return it branches. I don't mind rewriting it like you suggested. Which one do you prefer?

drew2a replied:
> I changed it to use a dictionary first to make it clearer, and second, to avoid branching. A series of if statements like you wrote is clearer than what I wrote, but in return it branches. I don't mind rewriting it like you suggested. Which one do you prefer?

I lean towards the version that incorporates branches. As far as I can see, there don't appear to be any disadvantages to using them.

> The first converts 0 => -1, -1 => 1. The second does so in reverse, meaning -1 => 0, 1 => -1. 
... 
Which option should I go with?

I believe option (1) seems to be the most straightforward and uncomplicated, so I would favor it over the other options.

---

Concerning the preference of using a dictionary over branching — we could consider seeking a second opinion on this matter.

@kozlovsky what version would you prefer?

1:
```python
    @staticmethod
    def convert_rate(rate):
        # Rate conversion due to the fact that we had a different system with Swift
        # and the old python BitTorrent core: unlimited == 0, stop == -1, else rate in kbytes
        if rate == -1:
            return 1
        if rate == 0:
            return -1
        return rate * 1024
```


2:
```python
    @staticmethod
    def convert_rate(rate):
        # Rate conversion due to the fact that we had a different system with Swift
        # and the old python BitTorrent core: unlimited == 0, stop == -1, else rate in kbytes
        torrent_rate = {0: -1, -1: 1}
        return torrent_rate.get(rate, rate * 1024)
```

kozlovsky replied:
Usually, I prefer the shorter option for simple expressions, but in this case, the first option with ifs looks clearer to me; it is easier to understand. So, I vote for the first option :)

Solomon1732 commented on src/tribler/core/components/libtorrent/download_manager/download_manager.py:
```diff
@@ -780,7 +784,7 @@ def update_trackers(self, infohash, trackers):
             old_trackers = old_def.get_trackers()
             new_trackers = list(set(trackers) - old_trackers)
-            all_trackers = list(old_trackers) + new_trackers
+            all_trackers = [*old_trackers, *new_trackers]
```
The original version creates an intermediary list. Mine doesn't 😅. If you prefer it, I'll remove this change.

drew2a replied:
I now understand your approach and must say that your version is indeed superior. Please proceed with it.

Solomon1732 commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -27,9 +28,10 @@ def create_fake_download_and_state():
-    fake_download.checkpoint = lambda: succeed(None)
-    fake_download.stop = lambda: succeed(None)
-    fake_download.shutdown = lambda: succeed(None)
+    succeed_none = functools.partial(succeed, None)
+    fake_download.checkpoint = succeed_none
+    fake_download.stop = succeed_none
+    fake_download.shutdown = succeed_none
```
I see. I'll change it accordingly.

Solomon1732 commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -138,9 +140,10 @@ async def test_get_metainfo_with_already_added_torrent(fake_dlmgr):
-    download_impl.checkpoint = lambda: succeed(None)
-    download_impl.stop = lambda: succeed(None)
-    download_impl.shutdown = lambda: succeed(None)
+    succeed_none = functools.partial(succeed, None)
+    download_impl.checkpoint = succeed_none
+    download_impl.stop = succeed_none
+    download_impl.shutdown = succeed_none
```
Same here: I'll change it 😃.

Solomon1732 commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -155,10 +158,10 @@ async def test_start_download_while_getting_metainfo(fake_dlmgr):
 
     metainfo_dl = MagicMock()
-    metainfo_dl.get_def = lambda: MagicMock(get_infohash=lambda: infohash)
+    metainfo_dl.get_def = functools.partial(MagicMock, get_infohash=lambda: infohash)
```
Oh, I see. I didn't notice. Thanks! I'll remove it.

Solomon1732 commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -181,7 +184,7 @@ async def test_start_download(fake_dlmgr):
 
     mock_handle = MagicMock()
-    mock_handle.info_hash = lambda: hexlify(infohash)
+    mock_handle.info_hash = functools.partial(hexlify, infohash)
```
Same as in https://github.com/Tribler/tribler/pull/7466#discussion_r1248052921.

Solomon1732 commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -241,7 +244,7 @@ async def test_start_download_existing_handle(fake_dlmgr):
 
     mock_handle = MagicMock()
-    mock_handle.info_hash = lambda: hexlify(infohash)
+    mock_handle.info_hash = functools.partial(hexlify, infohash)
```
Ditto.

Solomon1732 commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -406,7 +409,7 @@ def mocked_load_checkpoint(filename):
 
     mocked_load_checkpoint.called = False
-    fake_dlmgr.get_checkpoint_dir = lambda: Path(tmpdir)
+    fake_dlmgr.get_checkpoint_dir = functools.partial(Path, tmpdir)
```
Ditto.

Solomon1732 commented:
When this PR is ready to be pulled, I will of course rewrite commit messages and squash accordingly. Right now it's a bit of a mess since it's a work in progress.

Solomon1732 commented:
All those `*Mock` instantiations I did are going wrong somewhere and I'm not sure where 😓. I'd test it locally, except #7252 prevents me.

Solomon1732 commented:
Does [this](https://github.com/Tribler/tribler/actions/runs/5490101408/jobs/10005117226?pr=7466) coverage error relate to #7382? I'm asking since the latest force push from my fork only rebased from upsteam. I did no additional changes since the last one. Yet the test failed. Do my changes affect the failed file?
> /home/runner/work/_temp/7520c105-57cf-491b-9035-f33c995d6039.sh: line 1:  1875 Segmentation fault      (core dumped) coverage run --source=./src/tribler/core -p -m pytest ${PYTEST_CORE_ARGUMENTS_LINUX}
src/tribler/core/upgrade/tags_to_knowledge/tests/test_migration.py::test_migration 
Error: Process completed with exit code 139.

Solomon1732 commented:
This came from a recent pull request:
```py
def test_start_download_existing_download(fake_dlmgr):
    """
    Testing the addition of a torrent to the libtorrent manager, if there is a pre-existing download.
    """
    infohash = b'a' * 20

    mock_download = MagicMock(
        get_def=MagicMock(return_value=MagicMock(get_trackers=set))
    )
    mock_ltsession = MagicMock()

    fake_dlmgr.downloads[infohash] = mock_download
    fake_dlmgr.get_session = lambda *_: mock_ltsession

    download = await fake_dlmgr.start_download(tdef=TorrentDefNoMetainfo(infohash, 'name'), checkpoint_disabled=True)
    assert download == mock_download
    fake_dlmgr.downloads.clear()
```

Notice there's an `await` outside of an `async` context. It causes the test to fail.

qstokkink commented:
@Solomon1732 I see that this PR has been open for quite a while now. Do you need any help in getting this PR ready?

Solomon1732 commented:
> @Solomon1732 I see that this PR has been open for quite a while now. Do you need any help in getting this PR ready?

@qstokkink I admit I do. I don't know how to cover the following warnings in `codecove/patch` with tests:
- https://github.com/Tribler/tribler/pull/7466/files#annotation_16642491167
- https://github.com/Tribler/tribler/pull/7466/files#annotation_16642491170
- https://github.com/Tribler/tribler/pull/7466/files#annotation_16642491172
- https://github.com/Tribler/tribler/pull/7466/files#annotation_16642491174
- https://github.com/Tribler/tribler/pull/7466/files#annotation_16642491177
- https://github.com/Tribler/tribler/pull/7466/files#annotation_16642491181

You can see they're all code changes which are internal, and don't necessarily affect a function's output. I don't know how to approach them. If I can get ideas or pointers for how to write tests for these failures, I would be happy.

qstokkink commented:
Ah yes, those are pretty nasty. I could cover most of them (except the `dummy_mode` block in `download_manager.py`) as follows:

<details> <summary> Click me for changes! </summary>

```patch
diff --git a/src/tribler/core/components/libtorrent/tests/test_download_manager.py b/src/tribler/core/components/libtorrent/tests/test_download_manager.py
index 493ee66eb..e3b1a598b 100644
--- a/src/tribler/core/components/libtorrent/tests/test_download_manager.py
+++ b/src/tribler/core/components/libtorrent/tests/test_download_manager.py
@@ -1,6 +1,7 @@
 import asyncio
 from asyncio import Future, gather, get_event_loop, sleep
-from unittest.mock import MagicMock
+from functools import partial
+from unittest.mock import MagicMock, Mock
 from unittest.mock import AsyncMock
 
 import pytest
@@ -491,3 +492,71 @@ async def test_check_for_dht_ready(fake_dlmgr):
     fake_dlmgr.get_session().status().dht_nodes = 1000
     # If the session has enough peers, it should finish instantly
     await fake_dlmgr._check_dht_ready()
+
+
+def test_update_trackers(fake_dlmgr: DownloadManager) -> None:
+    fake_download, dl_state = create_fake_download_and_state()
+    fake_dlmgr.downloads[fake_download.infohash] = fake_download
+    fake_metainfo = {b'info': {b'name': b"test_download"}}
+    fake_download.get_def().metainfo = fake_metainfo
+
+    fake_tracker1 = "127.0.0.1/test-announce1"
+
+    fake_dlmgr.update_trackers(fake_download.infohash, [fake_tracker1])
+
+    assert fake_metainfo["announce"] == fake_tracker1
+    assert "announce-list" not in fake_metainfo
+
+
+def test_update_trackers_list(fake_dlmgr: DownloadManager) -> None:
+    fake_download, dl_state = create_fake_download_and_state()
+    fake_dlmgr.downloads[fake_download.infohash] = fake_download
+    fake_metainfo = {b'info': {b'name': b"test_download"}}
+    fake_download.get_def().metainfo = fake_metainfo
+
+    fake_tracker1 = "127.0.0.1/test-announce1"
+    fake_tracker2 = "127.0.0.1/test-announce2"
+
+    fake_dlmgr.update_trackers(fake_download.infohash, [fake_tracker1, fake_tracker2])
+
+    assert "announce" not in fake_metainfo
+    assert fake_metainfo["announce-list"] == [[fake_tracker1, fake_tracker2]]
+
+
+def test_update_trackers_list_append(fake_dlmgr: DownloadManager) -> None:
+    fake_download, dl_state = create_fake_download_and_state()
+    fake_dlmgr.downloads[fake_download.infohash] = fake_download
+    fake_metainfo = {b'info': {b'name': b"test_download"}}
+    fake_download.get_def().metainfo = fake_metainfo
+
+    fake_tracker1 = "127.0.0.1/test-announce1"
+    fake_tracker2 = "127.0.0.1/test-announce2"
+
+    fake_dlmgr.update_trackers(fake_download.infohash, [fake_tracker1])
+    fake_download.get_def().get_tracker = Mock(return_value=fake_tracker1)
+    fake_dlmgr.update_trackers(fake_download.infohash, [fake_tracker2])
+
+    assert fake_metainfo["announce"] == fake_tracker1
+    assert fake_metainfo["announce-list"] == [[fake_tracker1, fake_tracker2]]
+
+
+def test_get_download_rate_limit(fake_dlmgr: DownloadManager) -> None:
+    fake_value = 42
+    settings = {}
+    fake_dlmgr.get_session = Mock(return_value=Mock(get_settings=Mock(return_value=settings)))
+    fake_dlmgr.get_session().download_rate_limit = partial(settings.get, "download_rate_limit")
+    fake_dlmgr.ltsessions = {0: fake_dlmgr.get_session()}
+    fake_dlmgr.set_download_rate_limit(fake_value, 0)
+
+    assert fake_value == fake_dlmgr.get_download_rate_limit()
+
+
+def test_get_upload_rate_limit(fake_dlmgr: DownloadManager) -> None:
+    fake_value = 42
+    settings = {}
+    fake_dlmgr.get_session = Mock(return_value=Mock(get_settings=Mock(return_value=settings)))
+    fake_dlmgr.get_session().upload_rate_limit = partial(settings.get, "upload_rate_limit")
+    fake_dlmgr.ltsessions = {0: fake_dlmgr.get_session()}
+    fake_dlmgr.set_upload_rate_limit(fake_value, 0)
+
+    assert fake_value == fake_dlmgr.get_upload_rate_limit()
diff --git a/src/tribler/core/components/tunnel/tests/test_dispatcher.py b/src/tribler/core/components/tunnel/tests/test_dispatcher.py
index c31c5cea7..e068570a2 100644
--- a/src/tribler/core/components/tunnel/tests/test_dispatcher.py
+++ b/src/tribler/core/components/tunnel/tests/test_dispatcher.py
@@ -1,7 +1,8 @@
+import asyncio
 from unittest.mock import Mock
 
 import pytest
-from ipv8.messaging.anonymization.tunnel import CIRCUIT_STATE_EXTENDING, CIRCUIT_STATE_READY, CIRCUIT_TYPE_DATA
+from ipv8.messaging.anonymization.tunnel import CIRCUIT_STATE_EXTENDING, CIRCUIT_STATE_READY, CIRCUIT_TYPE_DATA, Circuit
 from ipv8.util import succeed
 
 from tribler.core.components.tunnel.community.dispatcher import TunnelDispatcher
@@ -127,3 +128,35 @@ def test_check_connections(dispatcher, mock_circuit):
     assert connection not in dispatcher.con_to_cir
     assert mock_circuit.circuit_id not in dispatcher.cid_to_con
     assert 2 in dispatcher.cid_to_con
+
+
+async def test_on_data_after_select(dispatcher: TunnelDispatcher, mock_circuit: Circuit) -> None:
+    mock_connection = Mock(udp_connection=None)
+    mock_request = Mock(destination=("0.0.0.0", 1024), data=b'a')
+    dispatcher.set_socks_servers([mock_connection.socksserver])
+    dispatcher.tunnels.circuits = {}
+    dispatcher.tunnels.create_circuit = Mock(return_value=mock_circuit)
+    dispatcher.on_socks5_udp_data = Mock(return_value=None)
+    mock_circuit.ready = asyncio.Future()
+
+    mock_circuit.ready.set_result(True)
+    dispatcher.select_circuit(mock_connection, mock_request)
+    await asyncio.sleep(0)
+
+    assert dispatcher.on_socks5_udp_data.called
+
+
+async def test_on_data_after_select_no_result(dispatcher: TunnelDispatcher, mock_circuit: Circuit) -> None:
+    mock_connection = Mock(udp_connection=None)
+    mock_request = Mock(destination=("0.0.0.0", 1024), data=b'a')
+    dispatcher.set_socks_servers([mock_connection.socksserver])
+    dispatcher.tunnels.circuits = {}
+    dispatcher.tunnels.create_circuit = Mock(return_value=mock_circuit)
+    dispatcher.on_socks5_udp_data = Mock(return_value=None)
+    mock_circuit.ready = asyncio.Future()
+
+    mock_circuit.ready.set_result(None)
+    dispatcher.select_circuit(mock_connection, mock_request)
+    await asyncio.sleep(0)
+
+    assert not dispatcher.on_socks5_udp_data.called
```

</details>

The "dummy mode" code seems like dead code to me (and this should not be in our production code at all IMO).

Solomon1732 commented:
@qstokkink Thanks a bunch! I'll work on adding your changes to my PR. I'll also remove `dummy_mode` while I'm at it.

qstokkink commented:
@Solomon1732 No worries, I'm happy to help.

I would advise against removing `dummy_mode` in this PR though. Our testing infrastructure is complex, to say the least. You may end up with really nasty and annoying refactoring. Perhaps this would be better suited to a follow-up PR.

drew2a commented:
@qstokkink @Solomon1732, we can remove all `dummy_mode` from our code. It seems that it should be done in a separate PR, but that could lead to annoying rebasing for @Solomon1732 with the current PR.

As an option: we can accept all uncovered lines in this PR and remove `dummy_mode` in the next one.


<img width="600" alt="image" src="https://github.com/Tribler/tribler/assets/13798583/8071f95c-7c6a-4fc2-a3cf-a3ae5d1d6f34">

drew2a commented:
@Solomon1732 @qstokkink, as an option, we can ignore all uncovered lines related to `dummy_mode` and remove/fix them in the next one.

@qstokkink, could you please create an issue regarding the potential removal of `dummy_mode`? I agree with your suggestions, but we should analyze how we use `dummy_mode` and whether we can safely remove it (I can perform the analysis).

qstokkink commented:
@drew2a sure 👍

Solomon1732 commented:
@drew2a will do as you advise. Thanks a bunch! Yeah, it looks like it'll make this PR a hairy one.

qstokkink commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -455,3 +492,71 @@ async def test_check_for_dht_ready(fake_dlmgr):
+    fake_dlmgr.update_trackers(fake_download.infohash, [fake_tracker1, fake_tracker2])
+
+    assert "announce" not in fake_metainfo
+    assert fake_metainfo["announce-list"] == [[fake_tracker1, fake_tracker2]]
```
Apologies, my bad: on different operating systems (win + mac apparently) the set unwrapping is different.

The output can be either `[[fake_tracker1, fake_tracker2]]` **or** `[[fake_tracker2, fake_tracker1]]`. The same is true for the next test.

Solomon1732 replied:
Is the following code acceptable?
```py
assert (
    fake_metainfo["announce-list"] == [[fake_tracker1, fake_tracker2]] or
    fake_metainfo["announce-list"] == [[fake_tracker2, fake_tracker1]]
)
```

qstokkink replied:
Looks good to me 👍

Solomon1732 replied:
Which option? I added one in an edit to my message above.

qstokkink replied:
Ah, I see. 👀 I was referring to the original code you posted with the `or`: the first option. I'd rather avoid checking the Operating System (unless absolutely no other options exist).

Solomon1732 replied:
Will do. Thanks 😃

Solomon1732 commented:
I have no idea how to fix the errors "[final newline missing](https://github.com/Tribler/tribler/pull/7466/files#annotation_16768034397)" and "[Unexpected line ending format. There is 'CRLF' while it should be 'LF'](https://github.com/Tribler/tribler/pull/7466/files#annotation_16768034406)." I followed both Codasity's article [Error caused by incompatible line endings](https://docs.codacy.com/faq/troubleshooting/error-line-endings/) and Github's page [Configuring Git to handle line endings](https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings) but to no avail. I can't find any CRLF ending in any python file in the repo of my PR.

qstokkink commented:
@Solomon1732 are you using PyCharm? If so, there is a sneaky setting in the bottom right:

![settings](https://github.com/Tribler/tribler/assets/3630389/fc0b8136-fbca-4829-b846-09d4c95f913e)

Solomon1732 commented:
@qstokkink I'm using Visual Studio Code and Geany if I find VSC too cumbersome for this or that task.

qstokkink commented:
@Solomon1732 IIRC VSC also has its line endings setting in the bottom-right corner somewhere.

Solomon1732 commented:
I finally fixed it somehow. I did an ugly hack of committing, squashing, and force pushing. It should look fine on the PR, though.

qstokkink commented:
👍 whatever works. It seems our Codacy check did not run (for some reason) now though.

Solomon1732 commented:
Oh. Should the tests be reran, then? To ensure Codasity reports a solid result on this matter?

drew2a commented on src/tribler/core/components/libtorrent/download_manager/download_manager.py:
```diff
@@ -363,7 +389,7 @@ def set_max_connections(self, conns, hops=None):
         # Rate conversion due to the fact that we had a different system with Swift
         # and the old python BitTorrent core: unlimited == 0, stop == -1, else rate in kbytes
-        libtorrent_rate = int(-1 if rate == 0 else (1 if rate == -1 else rate * 1024))
+        libtorrent_rate = self.__class__.convert_rate(rate=rate)
```
This version is a bit easier to read:

```suggestion
        libtorrent_rate = self.convert_rate(rate=rate)
```

drew2a commented on src/tribler/core/components/libtorrent/download_manager/download_manager.py:
```diff
@@ -374,19 +400,19 @@ def get_upload_rate_limit(self, hops=None):
         # and the old python BitTorrent core: unlimited == 0, stop == -1, else rate in kbytes
         libtorrent_rate = self.get_session(hops).upload_rate_limit()
-        return 0 if libtorrent_rate == -1 else (-1 if libtorrent_rate == 1 else libtorrent_rate / 1024)
+        return self.__class__.reverse_convert_rate(rate=libtorrent_rate)
```
This version is a bit easier to read:

```suggestion
        return self.reverse_convert_rate(rate=libtorrent_rate)
```

drew2a commented on src/tribler/core/components/libtorrent/download_manager/download_manager.py:
```diff
@@ -374,19 +400,19 @@ def get_upload_rate_limit(self, hops=None):
 
     def set_download_rate_limit(self, rate, hops=None):
-        libtorrent_rate = int(-1 if rate == 0 else (1 if rate == -1 else rate * 1024))
+        libtorrent_rate = self.__class__.convert_rate(rate=rate)
```
This version is a bit easier to read:

```suggestion
        libtorrent_rate = self.convert_rate(rate=rate)
```

drew2a commented on src/tribler/core/components/libtorrent/download_manager/download_manager.py:
```diff
@@ -374,19 +400,19 @@ def get_upload_rate_limit(self, hops=None):
-        libtorrent_rate = self.get_session(hops).download_rate_limit()
-        return 0 if libtorrent_rate == -1 else (-1 if libtorrent_rate == 1 else libtorrent_rate / 1024)
+        libtorrent_rate = self.get_session(hops=hops).download_rate_limit()
+        return self.__class__.reverse_convert_rate(rate=libtorrent_rate)
```
This version is a bit easier to read:

```suggestion
        return self.reverse_convert_rate(rate=libtorrent_rate)
```

drew2a commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -22,24 +24,26 @@ def create_fake_download_and_state():
+        checkpoint=AsyncMock(),
+        stop=AsyncMock(),
+        shutdown=AsyncMock(),
+        config = MagicMock(
+           get_hops=MagicMock(return_value=0),
+            get_safe_seeding = MagicMock(return_value=True),
+        ),
+    )
+    fake_download.get_def().get_name_as_unicode = MagicMock(return_value="test.iso")
+    dl_state = MagicMock(
+        get_infohash = MagicMock(return_value=b'aaaa'),
+        get_status = MagicMock(return_value=DownloadStatus.SEEDING),
+        get_download = MagicMock(return_value=fake_download),
+    )
```
Please, fix the formatting:

```suggestion
        config=MagicMock(
            get_hops=MagicMock(return_value=0),
            get_safe_seeding=MagicMock(return_value=True),
        ),
    )
    fake_download.get_def().get_name_as_unicode = MagicMock(return_value="test.iso")
    dl_state = MagicMock(
        get_infohash=MagicMock(return_value=b'aaaa'),
        get_status=MagicMock(return_value=DownloadStatus.SEEDING),
        get_download=MagicMock(return_value=fake_download),
    )
```

drew2a commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -256,15 +275,29 @@ async def test_start_download_existing_handle(fake_dlmgr):
     await download.shutdown()
 
 
+def test_convert_rate():
+    assert DownloadManager.convert_rate(0) == -1  # 0 is a special value
+    assert DownloadManager.convert_rate(-1) == 1  # -1 is a special value
+    for test_num in [1, 2, -2]:
+        assert DownloadManager.convert_rate(test_num) == test_num * 1024
+
+
+def test_reverse_convert_rate():
+    assert DownloadManager.reverse_convert_rate(-1) == 0  # -1 is a special value
+    assert DownloadManager.reverse_convert_rate(1) == -1  # 1 is a special value
+    for test_num in [0, 2, -2]:
+        assert DownloadManager.reverse_convert_rate(test_num) == test_num // 1024
```
NIT: The tests have been written correctly, but their logic replicates the method's logic, which seems a bit unsafe to me (it feels like you're testing the methods using the method's logic).

I would suggest rewriting them as follows (by doing this rewriting I've found a bug in the second test):
```suggestion
def test_convert_rate():
    assert DownloadManager.convert_rate(0) == -1  # 0 is a special value
    assert DownloadManager.convert_rate(-1) == 1  # -1 is a special value

    assert DownloadManager.convert_rate(1) == 1024
    assert DownloadManager.convert_rate(2) == 2048
    assert DownloadManager.convert_rate(-2) == -2048


def test_reverse_convert_rate():
    assert DownloadManager.reverse_convert_rate(-1) == 0  # -1 is a special value
    assert DownloadManager.reverse_convert_rate(1) == -1  # 1 is a special value

    assert DownloadManager.reverse_convert_rate(0) == 0
    assert DownloadManager.reverse_convert_rate(1024) == 1
    assert DownloadManager.reverse_convert_rate(-2048) == -2
```

drew2a commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -455,3 +492,79 @@ async def test_check_for_dht_ready(fake_dlmgr):
+
+    assert "announce" not in fake_metainfo
+    # The order of the results changes between OSes, so this is needed
+    assert (
+        fake_metainfo["announce-list"] == [[fake_tracker1, fake_tracker2]] or
+        fake_metainfo["announce-list"] == [[fake_tracker2, fake_tracker1]]
+    )
```
Using the approach with the set looks a bit more correct and generic:

```suggestion
    tracker_list = fake_metainfo["announce-list"]
    actual_trackers = set(itertools.chain.from_iterable(tracker_list))
    assert actual_trackers == {fake_tracker1, fake_tracker2}
```

drew2a commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -455,3 +492,79 @@ async def test_check_for_dht_ready(fake_dlmgr):
+
+    assert fake_metainfo["announce"] == fake_tracker1
+    # The order of the results changes between OSes, so this is needed
+    assert (
+        fake_metainfo["announce-list"] == [[fake_tracker1, fake_tracker2]] or
+        fake_metainfo["announce-list"] == [[fake_tracker2, fake_tracker1]]
+    )
```
Using the approach with the set looks a bit more correct and generic:

```suggestion
    tracker_list = fake_metainfo["announce-list"]
    actual_trackers = set(itertools.chain.from_iterable(tracker_list))
    assert actual_trackers == {fake_tracker1, fake_tracker2}
```

drew2a commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -455,3 +492,79 @@ async def test_check_for_dht_ready(fake_dlmgr):
+    fake_dlmgr.ltsessions = {0: fake_dlmgr.get_session()}
+    fake_dlmgr.set_download_rate_limit(fake_value, 0)
+
+    assert fake_value == fake_dlmgr.get_download_rate_limit()
```
For asserts it is common practice to use the following syntax `assert actual==expected`
```suggestion
    assert fake_dlmgr.get_download_rate_limit() ==  fake_value
```

drew2a commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -455,3 +492,79 @@ async def test_check_for_dht_ready(fake_dlmgr):
+    fake_dlmgr.ltsessions = {0: fake_dlmgr.get_session()}
+    fake_dlmgr.set_upload_rate_limit(fake_value, 0)
+
+    assert fake_value == fake_dlmgr.get_upload_rate_limit()
```
For asserts it is common practice to use the following syntax `assert actual==expected`
```suggestion
    assert fake_dlmgr.get_upload_rate_limit()==fake_value
```

drew2a reviewed (changes requested):
Great work! I've added some comments, most of which are related to coding style. All of them come with suggestions, so if you agree with them, you could just apply the working version.

Solomon1732 commented:
@drew2a thanks! I applied all of your recommendations, so I'll mark all of them as "Resolved" instead of writing "done" and resolving each one.

Solomon1732 commented:
I suspect Codasity doesn't understand Python's `@pytest.fixture`. This is why it complains about "Redefining name 'fake_dlmgr' from outer scope (line 50)" in this PR for tests I added in `test_dowload_manager.py`.

drew2a commented:
@Solomon1732 

>I suspect Codasity doesn't understand Python's @pytest.fixture. This is why it complains about "Redefining name 'fake_dlmgr' from outer scope (line 50)" in this PR for tests I added in test_dowload_manager.py.

Don't go down this rabbit hole (see our discussion with @qstokkink https://github.com/Tribler/tribler/pull/7764#pullrequestreview-1774459402) :)

Basically, you have two options:
1. Suppress the warning.
2. Follow @qstokkink's suggestion.

I'll approve the PR in both cases, as both options are correct. However, @qstokkink most likely will not approve the PR if you suppress the warning. So, my advice is to opt for number 2 and follow @qstokkink's suggestion.

drew2a commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -22,24 +25,26 @@ def create_fake_download_and_state():
-
-    return fake_download, dl_state
+    fake_download = MagicMock(
+        get_def = MagicMock(return_value=tdef),
+        get_peerlist = MagicMock(return_value=[fake_peer]),
+        hidden = False,
+        checkpoint = AsyncMock(),
+        stop = AsyncMock(),
+        shutdown = AsyncMock(),
+        config = MagicMock(
+           get_hops = MagicMock(return_value=0),
+            get_safe_seeding = MagicMock(return_value=True),
+        ),
+    )
+    fake_download.get_def().get_name_as_unicode = MagicMock(return_value="test.iso")
+    dl_state = MagicMock(
+        get_infohash = MagicMock(return_value=b'aaaa'),
+        get_status = MagicMock(return_value=DownloadStatus.SEEDING),
+        get_download = MagicMock(return_value=fake_download),
```
Please, fix the formatting:
```suggestion
        get_def=MagicMock(return_value=tdef),
        get_peerlist=MagicMock(return_value=[fake_peer]),
        hidden=False,
        checkpoint=AsyncMock(),
        stop=AsyncMock(),
        shutdown=AsyncMock(),
        config=MagicMock(
            get_hops=MagicMock(return_value=0),
            get_safe_seeding=MagicMock(return_value=True),
        ),
    )
    fake_download.get_def().get_name_as_unicode = MagicMock(return_value="test.iso")
    dl_state = MagicMock(
        get_infohash=MagicMock(return_value=b'aaaa'),
        get_status=MagicMock(return_value=DownloadStatus.SEEDING),
        get_download=MagicMock(return_value=fake_download),
```

Solomon1732 replied:
Sorry! I missed it! Will do!

Solomon1732 replied:
Wait, even Black doesn't add spaces there. This is because these are parameters for constructors and functions. They're simply multi-lined. Do you still wish for me to space them?

Solomon1732 replied:
I'll go by the one you choose.

drew2a replied:
The option from Black looks similar to the one I provided you. So, feel free to use whichever one you prefer :)

Solomon1732 replied:
Yesterday I was tired and misread your answer. My bad 🙏 .

Solomon1732 commented:
>So, my advice is to opt for number 2 and follow @qstokkink's suggestion.
@drew2a I looked at it, and I will do it 😰 . I indeed don't want to go into that rabbit hole.

drew2a commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -256,15 +276,30 @@ async def test_start_download_existing_handle(fake_dlmgr):
+def test_reverse_convert_rate():
+    assert DownloadManager.reverse_convert_rate(-1) == 0  # -1 is a special value
+    assert DownloadManager.reverse_convert_rate(1) == -1  # 1 is a special value
+    for test_num in [0, 2, -2]:
+        assert DownloadManager.reverse_convert_rate(test_num) == test_num // 1024
```
There is a bug in this test. If you remove the loop and rewrite it with flat lines, you will see:

```python
def test_reverse_convert_rate():
    assert DownloadManager.reverse_convert_rate(-1) == 0  # -1 is a special value
    assert DownloadManager.reverse_convert_rate(1) == -1  # 1 is a special value

    assert DownloadManager.reverse_convert_rate(0) == 0
    assert DownloadManager.reverse_convert_rate(2) == 0
    assert DownloadManager.reverse_convert_rate(-2) == -1
```

So, please, use the previously suggested variant.

Solomon1732 replied:
I did locally. I must have pushed it wrong. Same for the message above. I'm not sure what I did wrong.

Solomon1732 replied:
No, wait, I did miss this one too. My bad.

Solomon1732 commented:
I'm not sure why this test failed, or if it's my PR's code or not. May I get some help? I think it may has to do with #7798?
```
Prepare all required actions
Run ./.github/actions/timeout
  with:
    command: python ./scripts/exit_node/run_exit_node.py --testnet --fragile
    duration: 10s
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.8.18/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.8.18/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.8.18/x64
    Python[2](https://github.com/Tribler/tribler/actions/runs/7424222588/job/20203375875?pr=7466#step:6:2)_ROOT_DIR: /opt/hostedtoolcache/Python/[3](https://github.com/Tribler/tribler/actions/runs/7424222588/job/20203375875?pr=7466#step:6:3).8.18/x6[4](https://github.com/Tribler/tribler/actions/runs/7424222588/job/20203375875?pr=7466#step:6:4)
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.8.18/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.8.18/x64/lib
    VIRTUAL_ENV: /home/runner/.virtualenvs/.venv
    PYTHONPATH: ./src:
    HELPER_INDEX: 1000
    HELPER_BASE: 2000
Run timeout --foreground -s SIGKILL 10s python ./scripts/exit_node/run_exit_node.py --testnet --fragile || code=$?; if [[ $code -ne 124 && $code -ne 0 ]]; then exit $code; fi
  timeout --foreground -s SIGKILL 10s python ./scripts/exit_node/run_exit_node.py --testnet --fragile || code=$?; if [[ $code -ne 124 && $code -ne 0 ]]; then exit $code; fi
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.8.18/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/3.8.18/x64/lib/pkgconfig
    Python_ROOT_DIR: /opt/hostedtoolcache/Python/3.8.18/x64
    Python2_ROOT_DIR: /opt/hostedtoolcache/Python/3.8.18/x64
    Python3_ROOT_DIR: /opt/hostedtoolcache/Python/3.8.18/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.8.18/x64/lib
    VIRTUAL_ENV: /home/runner/.virtualenvs/.venv
    PYTHONPATH: ./src:
    HELPER_INDEX: 1000
    HELPER_BASE: 2000
Traceback (most recent call last):
  File "./scripts/exit_node/run_exit_node.py", line 226, in <module>
    run(main())
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "./scripts/exit_node/run_exit_node.py", line 220, in main
    await service.start(args)
  File "./scripts/exit_node/run_exit_node.py", line 133, in start
    config = make_config(options)
  File "./scripts/exit_node/run_exit_node.py", line [5](https://github.com/Tribler/tribler/actions/runs/7424222588/job/20203375875?pr=7466#step:6:5)1, in make_config
    config.tunnel_community.competing_slots = options.competing_slots
  File "pydantic/main.py", line 35[7](https://github.com/Tribler/tribler/actions/runs/7424222588/job/20203375875?pr=7466#step:6:7), in pydantic.main.BaseModel.__setattr__
ValueError: "TunnelCommunitySettings" object has no field "competing_slots"
Error: Process completed with exit code 1.
```

qstokkink commented:
It seems the following two lines can/should be removed:

https://github.com/Tribler/tribler/blob/959169ed1c91290ea4d0caee6cc00d4b30bab3a7/scripts/exit_node/run_exit_node.py#L51

https://github.com/Tribler/tribler/blob/959169ed1c91290ea4d0caee6cc00d4b30bab3a7/scripts/exit_node/run_exit_node.py#L204

drew2a commented:
@Solomon1732 @qstokkink, I'll fix it within a couple of hours.

drew2a commented:
@Solomon1732 fixed.

drew2a commented on src/tribler/core/components/libtorrent/tests/test_download_manager.py:
```diff
@@ -238,39 +261,59 @@ async def test_start_download_existing_handle(fake_dlmgr):
+    assert DownloadManager.reverse_convert_rate(-1) == 0  # -1 is a special value
+    assert DownloadManager.reverse_convert_rate(1) == -1  # 1 is a special value
+
+    assert DownloadManager.reverse_convert_rate(0) == 0
+    assert DownloadManager.reverse_convert_rate(2) == 0
+    assert DownloadManager.reverse_convert_rate(-2) == -1
```
It's the same issue again. If you had applied the suggested solution from my original comment, it would have been fixed.

Please consider using numbers that actually represent the `reverse_convert_rate`, e.g.:

```python
    assert DownloadManager.reverse_convert_rate(1024) == 1
    assert DownloadManager.reverse_convert_rate(2048) == 2
    assert DownloadManager.reverse_convert_rate(-2048) == -2
```

Solomon1732 replied:
@drew2a This?
```py
    for test_num in [0, 2, -2]:
        assert DownloadManager.reverse_convert_rate(test_num) == test_num // 1024
```
If so, I misread your previous review. My bad.

Solomon1732 replied:
Found it:
```py
def test_convert_rate():
    assert DownloadManager.convert_rate(0) == -1  # 0 is a special value
    assert DownloadManager.convert_rate(-1) == 1  # -1 is a special value
    assert DownloadManager.convert_rate(1) == 1024
    assert DownloadManager.convert_rate(2) == 2048
    assert DownloadManager.convert_rate(-2) == -2048


def test_reverse_convert_rate():
    assert DownloadManager.reverse_convert_rate(-1) == 0  # -1 is a special value
    assert DownloadManager.reverse_convert_rate(1) == -1  # 1 is a special value
    assert DownloadManager.reverse_convert_rate(0) == 0
    assert DownloadManager.reverse_convert_rate(1024) == 1
    assert DownloadManager.reverse_convert_rate(-2048) == -2
```
Will fix them.

Solomon1732 replied:
Done.

drew2a reviewed (changes requested):
Looks good except for one small mistake in the tests, which can be easily fixed. I proposed a complete solution during the previous review round, but it appears to have been overlooked.

drew2a reviewed (approved):
Thank you for your patience.

drew2a commented:
It could be merged if @qstokkink or @kozlovsky don't have any objections 👍 .
